examples/pokemon-service/src/authz.rs

```diff
@@ -0,0 +1,176 @@
+            AuthorizeServiceError::AuthorizeError { message } => http::Response::builder()
+                .status(http::StatusCode::UNAUTHORIZED)
+                .body(aws_smithy_http_server::body::to_boxed(message))
+                .expect("attempted to build an invalid HTTP response; please file a bug report"),
+        }
+    }
+}
+
+macro_rules! impl_service {
+    ($($var:ident),*) => {
+        impl<S, Op, $($var,)*> Service<(Op::Input, ($($var,)*))> for AuthorizeService<Op, S>
+        where
+            Op: OperationShape,
+            S: Service<(Op::Input, ($($var,)*)), Error = Op::Error> + Clone + Send + 'static,
+            S::Future: Send,
+            Op::Input: Send + 'static,
+            $($var: Send + 'static,)*
+        {
+            type Response = S::Response;
+            type Error = AuthorizeServiceError<Op::Error>;
+            type Future =
+                Pin<Box<dyn std::future::Future<Output = Result<S::Response, Self::Error>> + Send>>;
+
+            fn poll_ready(
+                &mut self,
+                cx: &mut std::task::Context<'_>,
+            ) -> std::task::Poll<Result<(), Self::Error>> {
+                self.inner
+                    .poll_ready(cx)
+                    .map_err(|e| Self::Error::InnerServiceError(e))
+            }
+
+            fn call(&mut self, req: (Op::Input, ($($var,)*))) -> Self::Future {
+                let (input, exts) = req;
+
+                // Replacing the service is necessary to avoid readiness problems.
+                // https://docs.rs/tower/latest/tower/trait.Service.html#be-careful-when-cloning-inner-services
+                let service = self.inner.clone();
+                let mut service = std::mem::replace(&mut self.inner, service);
+
+                let fut = async move {
+                    // You could imagine that here we would make a more complex async call to perform the
+                    // actual authorization.
+                    // We would likely need to add bounds on `Op::Input`, `Op::Error`, or on data
+                    // types the plugin would hold if we wanted to do anything useful.
+                    let is_authorized = true;
+                    if !is_authorized {
+                        return Err(Self::Error::AuthorizeError {
+                            message: "Not authorized!".to_owned(),
+                        });
+                    }
+
+                    service
+                        .call((input, exts))
+                        .await
+                        .map_err(|e| Self::Error::InnerServiceError(e))
+                };
+                Box::pin(fut)
+            }
+        }
+    };
+}
+
+// If we want our plugin to be as reusable as possible, the service it applies should work with
+// inner services (i.e. operation handlers) that take a variable number of parameters. A Rust macro
+// is helpful in providing those implementations concisely.
+
```

Comment:
Nit: Do you think it will make it clear to the users if we add a bit more on the reasoning for invoking the macro eight times? For example something like:

Each handler function registered using `builder_with_plugins` must accept the operation's input type (if there is one). Additionally, it can take up to 7 different parameters, each of which must implement the `FromRequest` trait. Internally, the system invokes a different Tower service depending on the number of parameters that a handler accepts. To ensure that this `AuthorizeService` works with any of those inner services, we must implement it to handle up to 7 different types. Therefore, we invoke the impl_service macro 8 times.

Reply:
I added your comment. Note it's `FromParts` though: https://docs.rs/aws-smithy-http-server/latest/aws_smithy_http_server/request/trait.FromParts.html